src/android/Capture.java

```diff
@@ -287,20 +291,59 @@ private static void createWritableFile(File file) throws IOException {
         file.setWritable(true, false);
     }
 
+    private String[] getRequiredPermissions(String[] allPerms) {
+        // not allowed to do multiple individual permissions requests so returning an array
+        ArrayList<String> neededPerms = new ArrayList<String>(allPerms.length);
+        for (String perm : allPerms) {
+            if (!PermissionHelper.hasPermission(this, perm)) {
+                neededPerms.add(perm);
+            }
+        }
+        return neededPerms.toArray(new String[]{});
+    }
+
     /**
      * Sets up an intent to capture video.  Result handled by onActivityResult()
      */
     private void captureVideo(Request req) {
-        if(cameraPermissionInManifest && !PermissionHelper.hasPermission(this, Manifest.permission.CAMERA)) {
-            PermissionHelper.requestPermission(this, req.requestCode, Manifest.permission.CAMERA);
+        String[] neededPermissions = this.getRequiredPermissions(new String[] {
+                Manifest.permission.READ_EXTERNAL_STORAGE,
+                Manifest.permission.CAMERA,
+                Manifest.permission.WRITE_EXTERNAL_STORAGE});
+        if (neededPermissions.length > 0) {
+            PermissionHelper.requestPermissions(this, req.requestCode, neededPermissions);
         } else {
             Intent intent = new Intent(android.provider.MediaStore.ACTION_VIDEO_CAPTURE);
 
-            if(Build.VERSION.SDK_INT > 7){
-                intent.putExtra("android.intent.extra.durationLimit", req.duration);
-                intent.putExtra("android.intent.extra.videoQuality", req.quality);
+            final ContentResolver contentResolver = this.cordova.getActivity().getContentResolver();
+            final ContentValues cv = new ContentValues();
+            cv.put(MediaStore.Images.Media.MIME_TYPE, VIDEO_MP4); // 3gp in some cases?
+            try {
+                cordova.getThreadPool().submit(new Runnable() {
+                    @Override
+                    public void run() {
+                        videoUri = contentResolver.insert(MediaStore.Video.Media.EXTERNAL_CONTENT_URI, cv);
+                        File file = webView.getResourceApi().mapUriToFile(videoUri);
+                        if (file != null && !file.getParentFile().exists()) {
+                            file.getParentFile().mkdirs();
+                        }
+                    }
+                }).get(); // get() blocks the thread
+
+                LOG.d(LOG_TAG, "Taking a video and saving to: " + videoUri.toString());
```

Comment:
if line 331 blocks the thread, maybe this log should be up in the runnable, or before the blocking call?

Reply:
I'll put it before